src/UserSpaceInstrumentation/AllocateInTracee.h

```diff
@@ -15,25 +15,63 @@
 
-// Allocate `size` bytes of memory in the tracee's address space using mmap. The memory will have
-// read, write, and execute permissions. The memory allocated will start at `address`. `address`
-// needs to be aligned to page boundaries. If the memory mapping can not be placed at `address` an
-// error is returned.
-// If `address` is zero the placement of memory will be arbitray (compare the documenation of mmap:
-// https://man7.org/linux/man-pages/man2/mmap.2.html). Assumes we are already attached to the tracee
-// `pid` e.g. using `AttachAndStopProcess`.
-[[nodiscard]] ErrorMessageOr<uint64_t> AllocateInTracee(pid_t pid, uint64_t address, uint64_t size);
-
-// Free address range previously allocated with AllocateInTracee using munmap.
-// Assumes we are already attached to the tracee `pid` e.g. using `AttachAndStopProcess`.
-[[nodiscard]] ErrorMessageOr<void> FreeInTracee(pid_t pid, uint64_t address, uint64_t size);
-
-// Same as `AllocateInTracee` above but wraps the memory in a unique_resource that handles
-// deallocating the memory. Note that we still need to be attached (or attached again) to the tracee
-// when the unique_resource goes out of scope.
-[[nodiscard]] ErrorMessageOr<orbit_base::unique_resource<uint64_t, std::function<void(uint64_t)>>>
-AllocateInTraceeAsUniqueResource(pid_t pid, uint64_t address, uint64_t size);
+// Represents a chunk of memory in the tracee. This class is movable but not copyable so it is a
+// unique handle to that memory in the tracee.
+// `MemoryInTracee` gets default constructed in `kInvalid` state. The first thing that needs to
+// happen is to actually allocate memory in the trace using `Allocate`. The memory gets allocated in
+// writable state. In case one wants to execute code in the segment it needs to be made executable
+// with `MakeMemoryExecutable` later.
+// `Free` deallocates the memory. Note that for each of the non-const member functions we need to
+// execute code in the tracee. So we need to be attached; the tracee needs to be stopped.
+class MemoryInTracee {
+ public:
+  MemoryInTracee() = default;
+  MemoryInTracee(const MemoryInTracee&) = delete;
+  MemoryInTracee(MemoryInTracee&&) = default;
+  MemoryInTracee& operator=(const MemoryInTracee&) = delete;
+  MemoryInTracee& operator=(MemoryInTracee&&) = default;
+
+  // Allocate `size` bytes of memory in the tracee's address space using mmap. The memory will have
+  // write permissions. The memory allocated will start at `address`. `address` needs to be aligned
+  // to page boundaries. If the memory mapping can not be placed at `address` an error is returned.
+  // If `address` is zero the placement of memory will be arbitray (compare the documenation of
+  // mmap: https://man7.org/linux/man-pages/man2/mmap.2.html). Assumes we are already attached to
+  // the tracee `pid` using `AttachAndStopProcess`.
+  [[nodiscard]] ErrorMessageOr<void> Allocate(pid_t pid, uint64_t address, uint64_t size);
```

Comment:
Is there ever a case when you need to call `Allocate` not right after the constructor? I guess not as the constructor doesn't do anything. It would seem like this should be done in the constructor following RAII. The fact that you start in `State::kInvalid` state and with `pid_ = 1` supports this, I don't think there's a need for an instance to ever be in that state.